### ISSUE TITLE
Feedback from Mozilla on the i18n API

### ISSUE THREAD

**Issue by zbraniecki** [browserext/browserext#54]
Several initial comments:

* Mozilla is investing quite heavily in Project Fluent [0] which we believe to be significantly more functional and better fitting for the Web Localization than the Chrome i18n module is.
* In the paradigm we propose, getMessage should not be a sync operation. Client side localization may use fallbacks in which case they should be loaded lazily and that in turn means that the API should be async. The good news is that in our approach most of the localization is declarative via DOM, rather than imperative via JS APIs.

I'd like to use this opportunity to reevaluate how l10n of web extensions should work rather than taking Chrome's json file format.

I believe that it should be possible to come up with an API that will handle Chrome's API and the Mozilla proposed API or be flexible in picking one.
But I am concerned about standardizing an API that we find quite lacking in terms of features, which will require upcoming localization APIs to introduce a parallel API namespace to "browser.i18n".

To be clear, I don't expect Mozilla to block on this, and if you'll push for the API in the current model, we'll probably implement it alongside our proposed API, but I believe that it'll be a missed opportunity.

If you have time to work with us on improving the L10n API, I'm very happy to get our perspective and counterproposal more formal.

[0] http://projectfluent.io/

**Comment by zbraniecki** [comment]
CC @stasm, @pike

**Comment by zbraniecki** [comment]
Just a minor update. @stasm pointed out that our paradigm does not, in fact, require async operations. We can work with a sync API at the cost of not providing runtime language fallback.

We believe the fallback ends up being a highly valuable part of the runtime flow, so we'd like to suggest having in mind that whatever methods we settle on, Mozilla's experience indicate that eventually we'll want to add an async version, and it will become the more common use case.

In the light of that, I'd like to propose that:

a) We make the API able to handle other formats than the current "message.json". In particular, we can offer our format - FTL ( https://github.com/projectfluent/fluent/tree/master/spec ) which we believe to provide much greater localization capabilities than `message.json` does.
b) Come up with an API that provides a sync method name to retrieve a formatted message (`getFormattedMessage`, `formatMessage` etc.), and can later be extended with a sister async method. I'd like to in particular try to avoid the awkwardness of node's `getX` and `getXSync` or `getX` and `getXAsync` pairs. I do not currently have a name pair in mind that would work well tho.

In result I imagine the API to work more or less like this:

message.ftl:
```ftl
hello-world = Hello, { $user }
```

```javascript
browser.i18n.formatMessage("hello-world")
```

all the linking and context creation *could* follow the current pattern of picking up the l10n resource automatically based on directory paths.

But, Fluent also recognizes so called `localization contexts` which are bundles of resources operating in a single context. We imagine that such a feature could be also allowed, and could work like this:

```json
{
  "l10n": {
    "main": ["brand", "menu", "toolbar"],
    "extra": ["brand", "extra"]
  }
}
```

```
_locales/en/brand.ftl
_locales/en/menu.ftl
_locales/en/toolbar.ftl
_locales/en/extra.ftl
```
brand.ftl:
```
brandName = Avilo
```

extra.ftl:
```
debug-button = Debug { brandName }
```

```javascript
const l10nCtx = browser.i18n.getContext('extra');
l10nCtx.formatMessage("extra-message"); // "Debug Avilo"
```

**Comment by dead-claudia** [comment]
@zbraniecki The URL you have in the description 404s on me. I think [this](http://projectfluent.org/) is the correct URL, though, if you want to update it.

**Comment by zbraniecki** [comment]
Yep, .io domain registrar messed up and we had to move the domain. Sorry for the trouble!
